ce/api/util.py

```diff
@@ -108,42 +109,53 @@ def mean_datetime(datetimes):
     return datetime.fromtimestamp(mean, tz=timezone.utc)
 
 
+# valid cell method parameters for the API. Add new ones here as needed.
+VALID_CELL_METHOD_PARAMETERS = ("mean", "standard_deviation", "percentile")
+
+
 def validate_cell_method(cell_method):
-    return cell_method in ("mean", "standard_deviation")
-
-
-def find_matching_cell_methods(cell_methods, target_method):
-    def filter_on_method(cell_method, target_method):
-        pattern = r"time:[a-z\s]*time:\s+{}\s+over\s+(days|years)".format(target_method)
-        return re.fullmatch(pattern, cell_method)
-
-    # Older data sets were all climatological means and therefore the
-    # cell_method attribute never had to specify that they were such.  With the
-    # introduction of climatological standard deviation data sets the
-    # cell_method usage had to be updated such that we could differentiate
-    # between the different the data operations.  Thus any cell_method that
-    # doesn't match the updated version is considered to be a climatological
-    # mean.
-    #
-    # The conventions that were followed to create these cell_method attributes
-    # can be found here:
-    # http://cfconventions.org/cf-conventions/cf-conventions.html#cell-methods
-    if target_method == "mean":
-        return [
-            cell_method
-            for cell_method in cell_methods
-            if filter_on_method(cell_method, target_method)
-            or (
-                not filter_on_method(cell_method, target_method)
-                and not filter_on_method(cell_method, "standard_deviation")
-            )
-        ]
+    """Validate the cell_method parameter supplied by caller"""
+    return cell_method in VALID_CELL_METHOD_PARAMETERS
+
+
+def check_cell_method(cell_methods, target_method, default_to_mean=True):
+    """Determines whether the final method in a cell methods string
+    (corresponding to a statistical aggregation) matches the target.
+    If default_to_mean is true, treats errors and unrecognized methods
+    as though they are climatological means. This compensates for some
+    noisy cell_methods attributes in our data, all of which are
+    climatological means.
+    """
+    parsed = parse(cell_methods)
+    if target_method == "mean" and default_to_mean:
+        # determine means by process of elimination
+        nonmeans = [m for m in VALID_CELL_METHOD_PARAMETERS if m != "mean"]
+        return not parsed or parsed[-1].method.name not in nonmeans
+    elif parsed:
+        return parsed[-1].method.name == target_method
```

Comment:
Some tests of this method would be helpful. I see we are testing it indirectly via the API tests, but I think it would be good to test the method in isolation.

Reply:
> What happens the cases that not parsed and not (target_method == mean and default_to_mean? Looks like it falls through the ifs and no result (actually, None) is returned.

There _is_ actually an `else` after the `elif` that handles that case, but github formatted the changes extremely weirdly and hid it.

Reply:
My bad. Sorry.